### ISSUE TITLE
ContinuousFactor should accept scipy.stats distributions

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#774]
Currently `ContinuousFactor` only accepts a function for the `pdf` argument. It should accept instances of distributions from `scipy.stats`.

API should look something like:

``` python
from pgmpy.factors.continuous import ContinuousFactor
from scipy.stats import multivariate_normal

normal = multivariate_normal(mean=[3, 4], cov=[[0.3, 2], [0.3, 5]])
phi_normal = ContinuousFactor(['x', 'y'], normal)
```

In case of strings:

```
phi_normal = ContinuousFactor(['X', 'Y'], `dirichlet`, alpha=[3, 4])
```

**Comment by kris-singh** [comment]
Hi would like to work on this. Can you guide me.
I was thinking of adding list of all the multivariate distribution in scipy.stats to the ContinousFactor File and definig init constructor as follows
def __init__(self, variables, pdf_name, pdf_param):
    if pdf_name in mul_distb_list:
        pass
    else:
        raise 'error'
the only point is in checking the parameter list for the scipy distbributions

**Comment by ankurankan** [comment]
@kris-singh Yeah you will need to do something like that. I would suggest to define a dict instead of a list so that if you get a string as the parameter you can match it directly to the class using the dict. Also you will need to modify some of the methods to the class because there would be efficient methods for doing operations in case of standard distributions.

**Comment by yashu-seth** [comment]
@ankurankan Are you thinking of having only scipy.stats distributions as valid parameters or also allow user defined pdfs?

**Comment by ankurankan** [comment]
@yashu-seth Both.

**Comment by kris-singh** [comment]
@ankurankan  
1.You will need to modify some of the methods to the class because there would be efficient methods for doing operations in case of standard distributions. Can you explain this.

**Comment by ankurankan** [comment]
@kris-singh So we want the `ContinuousFactor` to work with both user defined probability density functions (which we currently have) and standard probability densities (like multivariate gaussian which you need to add). For standard distributions you can find formulas for marginalization, reduce and product (these are the methods of `ContinuousFactor`). For example for a multivariate gaussian $ N([x_1, x_2] \mid [\mu_1, \mu_2], [[\sigma_{11}, \sigma_{12}][\sigma_{21} \sigma_{22}] $, the marginal distribution over $ x_1 $ is given by $ N([x_1] \mid [\mu_1], [\sigma_{11}] $. So, in this case you can compute the marginal distribution directly and you will need to have conditions in the marginalize etc methods to work differently for different types of distributions.

You should be able to find these formulas for different distributions online. 

I would suggest you to read Pattern Recognition and Machine Learning by Bishop Pages: 78-93 if you need more details about the multivariate distribution (will also help you in understanding for the general case). 

And if I didn't make any sense at all, send an email to the mailing list and I will reply there in detail as I am unable to use latex here.

**Comment by kris-singh** [comment]
I am doing something like this http://ideone.com/EKrDsh
Is there a better way to deal with isinstance(pdf,scipy.stats._multivariate.multivariate_normal_frozen)
I will make a pull request when i implement the marginalize and reduce for normal and dirichlet distribution.
Is reduce equivalent to the conditional distribution P(y|x).

**Comment by ankurankan** [comment]
@kris-singh Just create a pull request. I will put my comments there (will be easier to track). Yes reduce is equivalent to conditional distribution.

**Comment by kris-singh** [comment]
@ankurankan  sorry for the late reply some nosetest are failing i will create a pull request as soon as figure them out
the pdf method arguments are not consistent in JGD pdf argument should be an list,but for continous factors we are using list . I think we should make this consistent

**Comment by yashu-seth** [comment]
This depends on #782
